plugins/timertop/timermodel.cpp

```diff
@@ -27,190 +27,387 @@
 {
-    return s_timerModel != nullptr;
+    if (index.row() < m_sourceModel->rowCount()) {
+        const QModelIndex sourceIndex = m_sourceModel->index(index.row(), 0);
+        QObject * const timerObject = sourceIndex.data(ObjectModel::ObjectRole).value<QObject *>();
+        const TimerId id = TimerId(timerObject);
+        auto it = m_timersInfo.find(id);
+
+        if (it == m_timersInfo.end()) {
+            it = m_timersInfo.insert(id, TimerIdInfo(id));
+        }
+
+        if (it != m_timersInfo.end())
+            return &it.value();
+    } else {
+        if (index.row() < m_sourceModel->rowCount() + m_freeTimersInfo.count())
+            return &m_freeTimersInfo[index.row() - m_sourceModel->rowCount()];
+    }
+
+    return nullptr;
 }
 
-TimerModel *TimerModel::instance()
+void TimerModel::triggerPushChanges()
 {
-    if (!s_timerModel)
-        s_timerModel = new TimerModel;
-    Q_ASSERT(s_timerModel);
-    return s_timerModel;
+    if (!TimerModel::isInitialized())
+        return;
+
+    QMutexLocker locker(Probe::objectLock());
+
+    if (!s_timerModel->m_pushTimer) {
+        s_timerModel->m_pushTimer = QSharedPointer<QTimer>(new QTimer);
+        s_timerModel->m_pushTimer->setSingleShot(true);
+        s_timerModel->m_pushTimer->setInterval(5000);
+        s_timerModel->m_pushTimer->moveToThread(s_timerModel->thread());
+        connect(s_timerModel->m_pushTimer.data(), SIGNAL(timeout()),
+                         s_timerModel->m_pushTimer.data(), SLOT(pushChanges()), Qt::QueuedConnection);
+    }
+
+    if (!s_timerModel->m_pushTimer->isActive()) {
+        s_timerModel->m_pushTimerStart.invoke(s_timerModel->m_pushTimer.data(), Qt::QueuedConnection);
+    }
 }
 
-TimerInfoPtr TimerModel::findOrCreateFreeTimerInfo(int timerId)
+void TimerModel::pushChanges()
 {
-    // First, return the timer info if it already exists
-    foreach (const TimerInfoPtr &freeTimer, m_freeTimers) {
-        if (freeTimer->timerId() == timerId)
-            return freeTimer;
+    if (!TimerModel::isInitialized())
+        return;
+
+    QMutexLocker locker(Probe::objectLock());
+    TimerIdInfoHash infoHash;
+
+    infoHash.reserve(s_gatheredTimersData.count());
+    for (auto it = s_gatheredTimersData.begin(), end = s_gatheredTimersData.end(); it != end; ++it) {
+        TimerIdData &itInfo = it.value();
+
+        if (itInfo.changed) {
+            infoHash.insert(it.key(), itInfo.toInfo(it.key().type()));
+            itInfo.changed = false;
+        }
     }
+    infoHash.squeeze();
 
-    // Create a new free timer, and emit the correct update signals
-    TimerInfoPtr timerInfo(new TimerInfo(timerId));
-    beginInsertRows(QModelIndex(), rowCount(), rowCount());
-    m_freeTimers.append(timerInfo);
-    endInsertRows();
-    return timerInfo;
+    s_timerModel->m_applyChangesMethod.invoke(s_timerModel, Qt::QueuedConnection,
+                                              Q_ARG(GammaRay::TimerIdInfoHash, infoHash));
```

Comment:
Why the extra queued invocation here? pushChanges() is already in the thread the model belongs too, and called from the event loop, no?